pkg/rpm/database.go

```diff
@@ -0,0 +1,285 @@
+			dbFiles[name] = data
+		}
+	}
+	if len(dbFiles) == 0 {
+		// Not rpm database was found.
+		return nil, nil
+	}
+	// Write the database files to the filesystem.
+	dbDir, err := os.MkdirTemp("", "rpm")
+	if err != nil {
+		logrus.WithError(err).Error("could not create temporary folder for the rpm database")
+		return nil, commonerr.ErrFilesystem
+	}
+	defer func() {
+		// Remove temporary directory if returning on errors.
+		if err != nil {
+			_ = os.RemoveAll(dbDir)
+		}
+	}()
+	for name, data := range dbFiles {
+		dbFilename := filepath.Join(dbDir, name)
+		err = os.WriteFile(dbFilename, data.Contents, 0700)
+		if err != nil {
+			logrus.WithError(err).Error("failed to create rpm database file")
+			return nil, commonerr.ErrFilesystem
+		}
+	}
+	return &rpmDatabase{
+		dbPath: dbDir,
+	}, nil
+}
+
+// QueryAll starts a query for all packages in the RPM database, returns a query
+// iterator. A query is actually performed by an underlying rpm sub-process, and
+// return is retrieved after parsing its output and status.
+func (d *rpmDatabase) QueryAll(opts QueryOpts) *rpmDatabaseQuery {
+	queryFormat := queryFmt
+	if opts.Testing {
+		queryFormat = queryFmtTest
+	}
+	cmd := exec.Command(
+		"rpm",
+		`--dbpath`, d.dbPath,
+		`--query`,
+		`--all`,
+		`--queryformat`, queryFormat,
+	)
+	stdoutPipe, err := cmd.StdoutPipe()
+	if err != nil {
+		return &rpmDatabaseQuery{err: err}
+	}
+	var errBuffer bytes.Buffer
+	cmd.Stderr = &errBuffer
+	if err := cmd.Start(); err != nil {
+		utils.IgnoreError(stdoutPipe.Close)
+		return &rpmDatabaseQuery{err: err}
+	}
+	return &rpmDatabaseQuery{
+		rpmWait:    cmd.Wait,
+		rpmStderr:  &errBuffer,
+		rpmStdout:  stdoutPipe,
+		rpmScanner: bufio.NewScanner(stdoutPipe),
+	}
+}
+
+// ProvidesFile return true if a package provides the specified path in the RPM
+// database. If the path is relative, we assume its relative to the root
+// directory.
+func (d *rpmDatabase) ProvidesFile(path string) bool {
+	if !strings.HasPrefix(path, "/") {
+		path = "/" + path
+	}
+	cmd := exec.Command(
+		"rpm",
+		`--dbpath`, d.dbPath,
+		`--query`,
+		`--whatprovides`, path,
+	)
+	if err := cmd.Run(); err != nil {
+		// When rpm does not provide a file, the expected exit status is 1. On non-zero
+		// we always default to returning the file is not provided, but anything other
+		// than 1 is considered unexpected.
+		if exitError, ok := err.(*exec.ExitError); ok && exitError.ExitCode() != 1 {
+			logrus.WithError(err).Errorf(
+				"unexpected exit status when querying %s is provided by an RPM package", path)
+		}
+		return false
+	}
+	// The rpm exited properly, which implies the file IS provided by an RPM
+	// package.
+	return true
+}
+
+// Delete removes all the RPM database files.
+func (d *rpmDatabase) Delete() error {
+	return os.RemoveAll(d.dbPath)
+}
+
+// Next retrieves the next item in the query and make it available through the Package call.
+func (q *rpmDatabaseQuery) Next() bool {
```

Comment:
ah ok I can see why we hardcoded the formats here instead of having the user pass it as an argument: we assume a specific format and allowing this to be customizable is probably not worth the effort. That being said, I still thinking instead of passing in a `Testing` option, we should still have the user pass in the desired format. Perhaps we can just expose the formats we support and have the user choose which one. If we want to ensure someone does not just pass any random string, we can probably make a `type queryFormat string` and then `var QueryFormat queryFormat = "some string"` to force people to use `rpm.QueryFormat`

Also, instead of calling it `queryFmtTest` perhaps we can rename it to something which makes it clear it is the same as the other but missing the module stuff.

Reply:
> That being said, I still think instead of passing in a Testing option, we should still have the user pass in the desired format. Perhaps we can just expose the formats we support and have the user choose which one.

To what gain? I prefer being explicit about why we have two different formats (for testing) rather than supporting an interface "disguised" as generic but intended solely for testing. It's more confusing.

Maybe we should focus on fixing the underlying problem (rpm version during development/testing != production), capture it as a tech debt of Scanner (which can harm other aspects not only code clarity) instead of making the interface less clear to satisfy a well-known practice -- in this case, it is a conscious decision to not follow the recommendation of tainting code with testing flags.

Reply:
Until we fix the tech debt, that is.

Reply:
I can see your argument, and I'm willing to keep this as a temporary solution. Can you just comment on the `Test` option what the difference is, why it exists, and a ticket number for fixing the underlying problem(s)?

Reply:
I don't think we should tie the code to issue the tracking since it will be hard to keep them in sync if the ticket state changes. I'll mark it as a FIXME instead.